File: src/core/runner.interactive.ts
import { generateText, LanguageModel } from "ai";
import chalk from "chalk";
import readline from "readline";
import { jsonFormatForLog, logger } from "./logger";
import { getWrappedTools } from "./tool";
import { Agent } from "./types";

/**
 * Create a readline interface for user input
 */
function createReadlineInterface() {
  return readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
}

const shouldStop = (response: any) => {
  return (
    response?.toolResults?.length === 1 &&
    response?.toolResults[0]?.toolName === "stop"
  );
};

/**
 * Run the agent in interactive mode with the given configuration.
 * Maintains conversation context between prompts.
 */
export async function runInteractive(
  agent: Agent,
  model: LanguageModel,
  initialPrompt?: string
): Promise<void> {
  const { systemPrompt, tools } = agent;
  let messageHistory = "";

  console.log("\n🤖 Interactive Dev Agent Session");
  console.log("Type 'stop' to end the session\n");

  const rl = createReadlineInterface();

  try {
    // If initial prompt is provided, start with it
    if (initialPrompt) {
      console.log(`👤 Initial prompt: ${initialPrompt}\n`);
    }

    while (true) {
      // Get user input
      const currentPrompt =
        initialPrompt ||
        (await new Promise<string>((resolve) => {
          rl.question("👤 Enter your prompt: ", resolve);
        }));
      initialPrompt = undefined; // Clear initial prompt after first use

      // Check if user wants to stop
      if (currentPrompt.toLowerCase() === "stop") {
        console.log("\n👋 Ending session. Goodbye!");
        break;
      }

      // Skip empty prompts
      if (!currentPrompt.trim()) {
        console.log("❌ Please enter a valid prompt");
        continue;
      }

      console.log("\n"); // Add spacing before agent response

      try {
        // Start new conversation with message history
        if (!messageHistory) {
          logger.info(chalk.bold.cyan("🤖 Starting Dev Agent"));
          logger.info(chalk.bold.blue("📝 System prompt:"), systemPrompt);
          messageHistory = `System prompt: ${systemPrompt}`;
        }

        logger.info(chalk.bold.magenta("❓ User prompt:"), currentPrompt);

        let isComplete = false;
        let fullPrompt = `${messageHistory}\nUser prompt: ${currentPrompt}`;
        let additionalPrompt = "";

        while (!isComplete) {
          const response = await generateText({
            model,
            prompt: fullPrompt,
            tools: getWrappedTools(tools),
          });

          if (response.text) {
            additionalPrompt += `Agent: ${response.text}\n`;
            logger.info(
              chalk.blue.bold("🤖 Agent response:"),
              response.text.trim()
            );
          }

          if (response.toolResults.length > 0) {
            const toolResultsStr = response.toolResults.map((r) => ({
              toolName: r.toolName,
              args: r.args,
              result: r.result,
            }));
            additionalPrompt += `Tool results: ${JSON.stringify(
              toolResultsStr
            )}\n`;

            response.toolResults.forEach((r) => {
              logger.info(
                chalk.cyan.bold(`🛠️ Tool ${chalk.italic(r.toolName)}`),
                chalk.gray("called with args:"),
                jsonFormatForLog(r.args),
                chalk.gray("and returned result:"),
                jsonFormatForLog(r.result)
              );
            });

            // Check if the model wants to stop
            if (shouldStop(response)) {
              const stopResult = response.toolResults[0].result as {
                success: boolean;
                result: string;
              };
              logger.success(chalk.bold.green("✅ Task Complete"));
              logger.info(
                chalk.green(
                  `🎯 Success: ${
                    stopResult.success
                      ? chalk.bold.green("TRUE")
                      : chalk.bold.red("FALSE")
                  }`
                )
              );
              logger.info(chalk.green("📊 Result:"), stopResult.result);
              isComplete = true;
              break;
            }
          } else {
            // If no tool calls and we got a text response, consider it complete
            isComplete = true;
          }

          // Update the prompt with the latest interaction
          fullPrompt = `${fullPrompt}\n---\n${additionalPrompt}`;
        }

        // Update message history with the complete interaction
        messageHistory = fullPrompt;
      } catch (error) {
        logger.error(
          `❌ Error: ${error instanceof Error ? error.message : String(error)}`
        );
        // Reset message history on error to start fresh
        messageHistory = "";
      }

      console.log("\n"); // Add spacing after agent response
    }
  } catch (error) {
    if (error instanceof Error && error.name === "InterruptError") {
      console.log("\n\n👋 Session interrupted. Goodbye!");
    } else {
      throw error;
    }
  } finally {
    rl.close();
    logger.close();
  }
}

File: python_app/core/runner_interactive.py
import asyncio
from typing import List
from pydantic_ai import Agent
from pydantic_ai.models.gemini import GeminiModel
from core.agent_config import AgentConfig
from core.logger import logger
from pydantic_ai.messages import ModelMessage

async def run_interactive(agent_config: AgentConfig, model: GeminiModel, prompt: str = None) -> None:
    """Run the agent in interactive mode with the given configuration.
    Maintains conversation context between prompts using message history.
    """
    # Create agent for conversation management
    agent = Agent(
        model=model,
        system_prompt=agent_config.system_prompt,
        tools=agent_config.tools,
    )

    print("\n🤖 Interactive Dev Agent Session")
    print("Type 'stop' to end the session\n")

    try:
        # Initialize message history
        message_history: List[ModelMessage] = []
        
        # If initial prompt is provided, start with it
        if prompt:
            print(f"👤 Initial prompt: {prompt}\n")
        
        while True:
            # Get user input if no initial prompt, or for subsequent prompts
            current_prompt = prompt if prompt is not None else input("👤 Enter your prompt: ").strip()
            prompt = None  # Clear initial prompt after first use
            
            # Check if user wants to stop
            if current_prompt.lower() == 'stop':
                print("\n👋 Ending session. Goodbye!")
                break
            
            # Skip empty prompts
            if not current_prompt:
                print("❌ Please enter a valid prompt")
                continue
            
            print("\n") # Add spacing before agent response
            
            try:
                # Start new conversation with message history
                if not message_history:
                    logger.info("🤖 Starting Dev Agent")
                    logger.info("📝 System prompt:", agent_config.system_prompt)
                
                logger.info("❓ User prompt:", current_prompt)
                
                async with agent.iter(current_prompt, message_history=message_history) as conversation:
                    async for node in conversation:
                        if hasattr(node, 'model_response'):
                            for part in node.model_response.parts:
                                if part.part_kind == 'tool-call':
                                    logger.tool_called(part.tool_name, part.args)
                                elif part.part_kind == 'text':
                                    logger.info("🤖 Agent response:", part.content.strip())
                        elif hasattr(node, 'request'):
                            for part in node.request.parts:
                                if part.part_kind == 'tool-return':
                                    logger.tool_returned(part.tool_name, part.content)
                        elif hasattr(node, 'data'):
                            logger.success("✅ Task Complete")
                            logger.info("📊 Final Result:", node.data)

                    # Update message history with the conversation result
                    message_history = conversation.result.all_messages()

            except Exception as e:
                logger.error(f"❌ Error: {str(e)}")
                # Reset message history on error to start fresh
                message_history = []
            
            print("\n") # Add spacing after agent response

    except KeyboardInterrupt:
        print("\n\n👋 Session interrupted. Goodbye!")
    finally:
        logger.close() 
